osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/SetConfigurationObjectCommandExecutor.java

```diff
@@ -47,172 +31,35 @@
-
-        final byte[] newConfigurationObjectFlagsByteArray = this.configurationObjectHelperService
-                .toByteArray(configurationFlags);
-
-        return new BitString(newConfigurationObjectFlagsByteArray, 16);
-    }
-
-    private void mergeOldFlags(final ConfigurationObjectDto configurationObjectOnDevice,
-            final List<ConfigurationFlagDto> configurationFlags) {
-        if (configurationObjectOnDevice != null) {
-            for (final ConfigurationFlagDto configurationFlagOnDevice : configurationObjectOnDevice
-                    .getConfigurationFlags().getConfigurationFlag()) {
-                final ConfigurationFlagDto configurationFlag = this
-                        .getConfigurationFlag(configurationFlags, configurationFlagOnDevice.getConfigurationFlagType());
-                if (configurationFlag == null) {
-                    configurationFlags.add(configurationFlagOnDevice);
-                }
-            }
-        }
-    }
-
-    private List<ConfigurationFlagDto> getNewFlags(final ConfigurationObjectDto configurationObject) {
-        final List<ConfigurationFlagDto> configurationFlags = new ArrayList<>();
-        if (configurationObject.getConfigurationFlags() == null) {
-            return configurationFlags;
-        }
-        for (final ConfigurationFlagDto configurationFlag : configurationObject.getConfigurationFlags()
-                .getConfigurationFlag()) {
-            if (!this.isForbidden(configurationFlag.getConfigurationFlagType())) {
-                configurationFlags.add(configurationFlag);
-            }
-        }
-        return configurationFlags;
-    }
-
-    /**
-     * Check if the configuratioFlag is forbidden. Check is done against the
-     * list of forbidden flag types
-     *
-     * @param configurationFlagType
-     *         the flag to check
-     *
-     * @return true if the flag is forbidden, else false
-     */
-    private boolean isForbidden(final ConfigurationFlagTypeDto configurationFlagType) {
-        for (final ConfigurationFlagTypeDto forbiddenFlagType : FLAGS_TYPES_FORBIDDEN_TO_SET) {
-            if (forbiddenFlagType.equals(configurationFlagType)) {
-                return true;
-            }
-        }
-        return false;
-    }
-
-    private ConfigurationFlagDto getConfigurationFlag(final Collection<ConfigurationFlagDto> flags,
-            final ConfigurationFlagTypeDto flagType) {
-        for (final ConfigurationFlagDto configurationFlag : flags) {
-            if (configurationFlag.getConfigurationFlagType().equals(flagType)) {
-                return configurationFlag;
-            }
-        }
-        return null;
+            final ConfigurationObjectDto configurationToSet) throws ProtocolAdapterException {
+        final Protocol protocol = Protocol.forDevice(device);
+        final GetConfigurationObjectService getService = this.protocolServiceLookup.lookupGetService(protocol);
```

Comment:
consider removing these lookups by one method (or class for all command executors) that grabs the bean by name from the Spring ApplicationContext.

Reply:
I prefer looking up beans by contract than by name, which is what the ProtocolServiceLookup does

Reply:
you don't need to return the impl, just return the contract (interface/abstract class)

Reply:
autowire both implementations by the contract, based on the protocol use either the DSMR_4 or SMR_5 impl. The lookup, and the protocolservice classes can be removed because they are just a wrapper around the Spring context.